File: service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java
package com.epam.esm.service.impl;

import com.epam.esm.exception.DuplicateEntityException;
import com.epam.esm.exception.EntityNotFoundException;
import com.epam.esm.exception.InvalidDataException;
import com.epam.esm.model.entity.GiftCertificate;
import com.epam.esm.model.entity.util.QueryParameters;
import com.epam.esm.repository.GiftCertificateRepository;
import com.epam.esm.repository.TagRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.PageImpl;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static com.epam.esm.service.impl.util.Constants.*;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class GiftCertificateServiceImplTest {

    @Mock
    private GiftCertificateRepository giftCertificateRepository;

    @Mock
    private TagRepository tagRepository;

    @InjectMocks
    private GiftCertificateServiceImpl giftCertificateService;

    @Test
    void getAll() {
        List<GiftCertificate> expected = Arrays.asList(FIRST_TEST_GIFT_CERTIFICATE, SECOND_TEST_GIFT_CERTIFICATE, THIRD_TEST_GIFT_CERTIFICATE);
        when(giftCertificateRepository.findAll(PAGE)).thenReturn(new PageImpl<>(expected));
        List<GiftCertificate> actual = giftCertificateService.getAll(PAGE_NUM, PAGE_SIZE);

        assertEquals(actual, expected);
    }

    @Test
    void getGiftCertificatesByParameters() {
        QueryParameters parameters = new QueryParameters("1", "", new ArrayList<>(), null, null);
        List<GiftCertificate> actual = giftCertificateService.getGiftCertificatesByParameters(parameters, PAGE_NUM, PAGE_SIZE);

        assertEquals(new ArrayList<>(), actual);
    }


    @Test
    void getById() {
        when(giftCertificateRepository.findById(TEST_ID)).thenReturn(Optional.of(FIRST_TEST_GIFT_CERTIFICATE));
        GiftCertificate actual = giftCertificateService.getById(TEST_ID);
        assertEquals(FIRST_TEST_GIFT_CERTIFICATE, actual);
    }

    @Test
    void getByIdShouldThrowEntityNotFoundException() {
        when(giftCertificateRepository.findById(NOT_EXIST_ID)).thenThrow(new EntityNotFoundException());
        assertThrows(EntityNotFoundException.class, () -> giftCertificateService.getById(NOT_EXIST_ID));
    }


    @Test
    void create() {
        when(giftCertificateRepository.save(GIFT_CERTIFICATE_TO_CREATE)).thenReturn(GIFT_CERTIFICATE_TO_CREATE);
        when(giftCertificateRepository.findByName(GIFT_CERTIFICATE_TO_CREATE.getName())).thenReturn(Optional.empty())
                .thenReturn(Optional.of(GIFT_CERTIFICATE_TO_CREATE));
        GiftCertificate actual = giftCertificateService.create(GIFT_CERTIFICATE_TO_CREATE);
        assertEquals(GIFT_CERTIFICATE_TO_CREATE, actual);
    }

    @Test
    void createShouldThrowInvalidDataException() {
        when(giftCertificateRepository.save(INVALID_GIFT_CERTIFICATE)).thenThrow(new InvalidDataException());
        assertThrows(InvalidDataException.class, () -> giftCertificateService.create(INVALID_GIFT_CERTIFICATE));
    }


    @Test
    void createShouldThrowDuplicateEntityException() {
        when(giftCertificateRepository.findByName(FIRST_TEST_GIFT_CERTIFICATE.getName())).thenReturn(Optional.of(FIRST_TEST_GIFT_CERTIFICATE));
        assertThrows(DuplicateEntityException.class, () -> giftCertificateService.create(FIRST_TEST_GIFT_CERTIFICATE));
    }

    @Test
    void update() {
        when(giftCertificateRepository.save(UPDATED_GIFT_CERTIFICATE)).thenReturn(UPDATED_GIFT_CERTIFICATE);
        when(giftCertificateRepository.findById(UPDATED_GIFT_CERTIFICATE.getId())).thenReturn(Optional.of(FIRST_TEST_GIFT_CERTIFICATE));
        GiftCertificate actual = giftCertificateService.update(UPDATED_GIFT_CERTIFICATE);
        assertEquals(UPDATED_GIFT_CERTIFICATE, actual);
    }

    @Test
    void updateShouldThrowInvalidDataException() {
        when(giftCertificateRepository.findById(INVALID_GIFT_CERTIFICATE.getId())).thenReturn(Optional.of(INVALID_GIFT_CERTIFICATE));
        when(giftCertificateRepository.save(INVALID_GIFT_CERTIFICATE)).thenThrow(new InvalidDataException());
        assertThrows(InvalidDataException.class, () -> giftCertificateService.update(UPDATED_GIFT_CERTIFICATE));
    }


    @Test
    void deleteByIdShouldThrowEntityNotFoundException() {
        assertThrows(EntityNotFoundException.class, () -> giftCertificateService.deleteById(NOT_EXIST_ID));
    }
}

File: service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.kt
package com.epam.esm.service.impl

import com.epam.esm.repository.GiftCertificateRepository
import com.epam.esm.service.impl.util.Constants.FIRST_TEST_GIFT_CERTIFICATE
import com.epam.esm.service.impl.util.Constants.GIFT_CERTIFICATE_TO_CREATE
import com.epam.esm.service.impl.util.Constants.NOT_EXIST_ID
import com.epam.esm.service.impl.util.Constants.PAGE
import com.epam.esm.service.impl.util.Constants.PAGE_NUM
import com.epam.esm.service.impl.util.Constants.PAGE_SIZE
import com.epam.esm.service.impl.util.Constants.SECOND_TEST_GIFT_CERTIFICATE
import com.epam.esm.service.impl.util.Constants.TEST_ID
import com.epam.esm.service.impl.util.Constants.THIRD_TEST_GIFT_CERTIFICATE
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.dao.DuplicateKeyException
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.test.StepVerifier
import org.mockito.Mockito.`when` as whenever

@ExtendWith(MockitoExtension::class)
class GiftCertificateServiceImplTest {

    @Mock
    private lateinit var giftCertificateRepository: GiftCertificateRepository

    @InjectMocks
    private lateinit var giftCertificateService: GiftCertificateServiceImpl

    @Test
    fun getAll() {
        val expected = listOf(FIRST_TEST_GIFT_CERTIFICATE, SECOND_TEST_GIFT_CERTIFICATE, THIRD_TEST_GIFT_CERTIFICATE)
        whenever(giftCertificateRepository.findAll(PAGE)).thenReturn(Flux.fromIterable(expected))

        val actual = giftCertificateService.getAll(PAGE_NUM, PAGE_SIZE)

        StepVerifier.create(actual)
            .expectNextCount(expected.size.toLong())
            .verifyComplete()
    }

    @Test
    fun getById() {
        whenever(giftCertificateRepository.findById(TEST_ID)).thenReturn(Mono.just(FIRST_TEST_GIFT_CERTIFICATE))

        val actual = giftCertificateService.getById(TEST_ID)

        StepVerifier.create(actual)
            .expectNext(FIRST_TEST_GIFT_CERTIFICATE)
            .verifyComplete()
    }

    @Test
    fun getByIdShouldThrowEntityNotFoundException() {
        whenever(giftCertificateRepository.findById(NOT_EXIST_ID)).thenReturn(Mono.empty())

        val actual = giftCertificateService.getById(NOT_EXIST_ID)

        StepVerifier.create(actual)
            .expectNext()
            .verifyError()
    }

    @Test
    fun create() {
        whenever(giftCertificateRepository.save(GIFT_CERTIFICATE_TO_CREATE)).thenReturn(
            Mono.just(GIFT_CERTIFICATE_TO_CREATE)
        )

        val actual = giftCertificateService.create(GIFT_CERTIFICATE_TO_CREATE)

        StepVerifier.create(actual)
            .expectNext(GIFT_CERTIFICATE_TO_CREATE)
            .verifyComplete()
    }

    @Test
    fun createShouldThrowDuplicateEntityException() {
        whenever(giftCertificateRepository.save(GIFT_CERTIFICATE_TO_CREATE)).thenReturn(
            Mono.error(DuplicateKeyException(""))
        )

        val actual = giftCertificateService.create(GIFT_CERTIFICATE_TO_CREATE)

        StepVerifier.create(actual)
            .expectNext()
            .verifyError()
    }

    @Test
    fun update() {
        GIFT_CERTIFICATE_TO_CREATE.id = "1"
        whenever(giftCertificateRepository.findById(TEST_ID)).thenReturn(
            Mono.just(GIFT_CERTIFICATE_TO_CREATE)
        )
        whenever(giftCertificateRepository.save(GIFT_CERTIFICATE_TO_CREATE)).thenReturn(
            Mono.just(GIFT_CERTIFICATE_TO_CREATE)
        )

        val actual = giftCertificateService.update(GIFT_CERTIFICATE_TO_CREATE)

        StepVerifier.create(actual)
            .expectNext(GIFT_CERTIFICATE_TO_CREATE)
            .verifyComplete()
    }

    @Test
    fun updateShouldThrowEntityNotFoundException() {
        GIFT_CERTIFICATE_TO_CREATE.id = "1"
        whenever(giftCertificateRepository.findById(TEST_ID)).thenReturn(
            Mono.empty()
        )

        val actual = giftCertificateService.update(GIFT_CERTIFICATE_TO_CREATE)

        StepVerifier.create(actual)
            .expectNext()
            .verifyError()
    }

    @Test
    fun deleteByIdShouldThrowEntityNotFoundException() {
        whenever(giftCertificateRepository.findById(NOT_EXIST_ID)).thenReturn(Mono.empty())

        val actual = giftCertificateService.deleteById(NOT_EXIST_ID)

        StepVerifier.create(actual)
            .expectNext()
            .verifyError()
    }
}
